src/common/types/row/tuple_data_scatter_gather.cpp

```diff
@@ -1448,36 +1286,103 @@ static void TupleDataCollectionWithinCollectionGather(const TupleDataLayout &lay
 
 	// Recurse
 	D_ASSERT(child_functions.size() == 1);
 	const auto &child_function = child_functions[0];
 	child_function.function(layout, heap_locations, child_list_size_before, scan_sel, scan_count,
-	                        COLLECTION::GetEntry(target), target_sel, combined_list_vector,
+	                        ListVector::GetEntry(target), target_sel, combined_list_vector,
 	                        child_function.child_functions);
 }
 
+// Special case: We cant currently handle arrays, so we need to replace each array with a list of the same type.
+static LogicalType TypeReplaceArrayWithList(const LogicalType &type) {
+	switch (type.id()) {
+	case LogicalTypeId::ARRAY: {
+		return LogicalType::LIST(TypeReplaceArrayWithList(ArrayType::GetChildType(type)));
+	}
+	case LogicalTypeId::LIST:
+		return LogicalType::LIST(TypeReplaceArrayWithList(ListType::GetChildType(type)));
+	case LogicalTypeId::STRUCT: {
+		auto children = StructType::GetChildTypes(type);
+		for (auto &child : children) {
+			child.second = TypeReplaceArrayWithList(child.second);
+		}
+		return LogicalType::STRUCT(children);
+	}
+	case LogicalTypeId::MAP: {
+		auto key_type = TypeReplaceArrayWithList(MapType::KeyType(type));
+		auto value_type = TypeReplaceArrayWithList(MapType::ValueType(type));
+		return LogicalType::MAP(key_type, value_type);
+	}
+	case LogicalTypeId::UNION: {
+		auto children = UnionType::CopyMemberTypes(type);
+		for (auto &child : children) {
+			child.second = TypeReplaceArrayWithList(child.second);
+		}
+		return LogicalType::UNION(children);
+	}
+	default:
+		return type;
+	}
+}
+
+static bool TypeHasArray(const LogicalType &type) {
+	switch (type.id()) {
+	case LogicalTypeId::ARRAY:
+		return true;
+	case LogicalTypeId::LIST:
+		return TypeHasArray(ListType::GetChildType(type));
+	case LogicalTypeId::STRUCT: {
+		auto children = StructType::GetChildTypes(type);
+		for (auto &child : children) {
+			if (TypeHasArray(child.second)) {
+				return true;
+			}
+		}
+		return false;
+	}
+	case LogicalTypeId::MAP: {
+		return TypeHasArray(MapType::KeyType(type)) || TypeHasArray(MapType::ValueType(type));
+	}
+	case LogicalTypeId::UNION: {
+		auto children = UnionType::CopyMemberTypes(type);
+		for (auto &child : children) {
+			if (TypeHasArray(child.second)) {
+				return true;
+			}
+		}
+		return false;
+	}
+	default:
+		return false;
+	}
+}
+
+static void TupleDataCollectionGatherArrayAsList(const TupleDataLayout &layout, Vector &row_locations,
+                                                 const idx_t col_idx, const SelectionVector &scan_sel,
+                                                 const idx_t scan_count, Vector &target,
+                                                 const SelectionVector &target_sel, Vector &dummy_vector,
+                                                 const vector<TupleDataGatherFunction> &child_functions) {
+
+	Vector cast_vector(TypeReplaceArrayWithList(target.GetType()));
```

Comment:
Instead of computing `TypeReplaceArrayWithList` and allocating the `cast_vector` (which is potentially deeply nested) for every scatter/gather, would it be possible to cache the type/vector and re-use them?

I think that `TypeReplaceArrayWithList` then only need to be called when initializing the `TupleDataGatherFunction` (to get the scatter/gather for `LIST` instead), and when initializing `TupleDataChunkState` (the allocated `cast_vector` can live there).

This would already make scatter/gather much more efficient.

Reply:
Yes, I initially tried to cache it in the tuple layout. The problem is that we don't always call gather with a TupleDataChunkState readily available. (perfect_hash_join_executor, radix_partitioning, join_hashtable)

Reply:
1. `perfect_hash_join_executor.cpp` won't be performed if there is an `ARRAY`: `CheckForPerfectJoinOpt` in `plan_comparison_join.cpp
2. `radix_partitioning.cpp` only gathers a column of type `HASH`, so no problem there
3. `join_hashtable.cpp` this one is a problem

For (3.) we could consider making the cached type/vector an `optional_ptr`, so that scanning/appending is more efficient, but it can't be done for random gathers.

I understand this is a bugfix PR, so it's also OK if we leave this optimization for later.

Reply:
Cool! Yeah I think I can get something like that going for 3. Ill give it a shot!